text/2021-02-24-coprocessor-plugin.md

```diff
@@ -0,0 +1,161 @@
+# Coprocessor Plugin
+
+## Summary
+
+Add a general and pluggable coprocessor framework for RawKV mode.
+
+## Motivation
+
+TiKV is the storage component in the TiDB ecosystem, however, the distribution computation principle suggests that computation should be as close to the data source as possible. Therefore, TiKV has embedded a subset of the TiDB executor framework to push down some computation tasks when applicable.
+
+But TiKV's capability should be far beyond that, as many distributed components can be built on top of TiKV, such as cache, full text search engine, graph database and NoSQL database. And same as TiDB, these product will also like to push down specific computation to TiKV, which requires the coprocessor to be customizable, aka pluggable.
+
+For instance, a full-text seraching engine will persist the origin document and n-gram index on TiKV. It'll be a waste of resource if we read back and then update the index from a client. In contrary, the coprocessor plugin can generate the index from the origin document, and update the index inplace. What's more, the coprocessor plugin can perform index scan directly on TiKV.
+
+The goals of the coprocessor plugin are:
+
+- Do what client can do (on single region)
+- Provide more guarantee than client does on RawKV
+    - Raft transaction on RawKV
+- Easy to use
+    - Out of box
+    - Easy to deploy
+- Robust
+    - Easy to debug
+    - Log support, metrics support
+
+## Detailed design
+
+### Dynamic vs statically
+
+Generally, there are two strategies to build a plugin framework: dynamically and statically, which means to load the plugin on startup or to embed in the binary on compilation.
+
+![Plugin Arch](../media/plugin-arch.png)
+
+They have both pros and cons:
+
+| Static | Dynamic |
+| -- | -- |
+| ◯ High performance | X Relatively slower |
+| ◯ Easy to deploy | X Complexify the deploy process |
+| X Build the entire TiKV | ◯ Easy to build |
+| X Build very slow | ◯ Build fast |
+| X Hard to debug | ◯ Easy to debug |
+
+In this RFC, we'll only focus on the dynamic plugin framework that works in RawKV mode.
+
+### Plugin runtime
+
+The plugin runtime is a new component settling in `tikv::server::service::kv::Service`. It loads the dylib plugin, dispatches coprocessor request to the plugin, and proxy the API calls from plugin to the [`Storage`](https://tikv.github.io/doc/tikv/storage/struct.Storage.html).
```

Comment:
could you elaborate more about the implementation of the plugin runtime? for example:

- how to implement a plugin under this framework?
- how to compile a plugin to a dynamic library?
- how to load the dynamic plugin library to the tikv-server process?
- is it possible to upgrade the dynamic plugin to a newer version without restarting the tikv-process?

Reply:
> how to implement a plugin under this framework?
> how to compile a plugin to a dynamic library?
> how to load the dynamic plugin library to the tikv-server process?

These are the main research goal ongoing in the LFX program, I could update the RFC after we're satisfied by a solution. However, here is an [poc project](https://github.com/andylokandy/plugin) that may answers these question for now.

> is it possible to upgrade the dynamic plugin to a newer version without restarting the tikv-process?

This is a good question, I think it's viable but I've though about it before. I'd try to achieve that during the program.